src/routes/get-files.ts

```diff
@@ -0,0 +1,227 @@
+import {
+  asBoolean,
+  asMap,
+  asNumber,
+  asObject,
+  asOptional,
+  asString
+} from 'cleaners'
+import Router from 'express-promise-router'
+
+import config from '../../config.json'
+import { dataStore } from '../db'
+import {
+  ApiResponse,
+  asStoreDirectory,
+  asStoreDirectoryDocument,
+  asStoreFile,
+  asStoreFileDocument,
+  asStoreRepoDocument,
+  StoreDirectoryDocument,
+  StoreFile,
+  StoreFileDocument,
+  StoreRepoDocument
+} from '../types'
+import {
+  getNameFromPath,
+  getParentPathsOfPath,
+  makeApiClientError
+} from '../utils'
+
+const { MAX_PAGE_SIZE } = config
+
+type FilesGetBody = ReturnType<typeof asFilesGetBody>
+const asFilesGetBody = asObject({
+  repoId: asString,
+  ignoreTimestamps: asOptional(asBoolean),
+  paths: asMap(asNumber)
+})
+
+interface FilesGetResponseData {
+  total: number
+  paths: FilesGetResponsePaths
+}
+
+interface FilesGetResponsePaths {
+  [path: string]: FilesGetResponseFile | FilesGetResponseDirectory
+}
+
+interface FilesGetResponseFile extends StoreFile {
+  timestamp: number
+}
+
+interface FilesGetResponseDirectory {
+  files: {
+    [path: string]: number
+  }
+  timestamp: number
+}
+
+export const getFilesRouter = Router()
+
+getFilesRouter.post('/get-files', async (req, res) => {
+  let body: FilesGetBody
+  let repoId: string
+  let ignoreTimestamps: boolean = false
+
+  // Validate request body
+  try {
+    body = asFilesGetBody(req.body)
+    repoId = body.repoId
+
+    if (body.ignoreTimestamps !== undefined) {
+      ignoreTimestamps = body.ignoreTimestamps
+    }
+
+    if (repoId === '') {
+      throw new Error(`Missing repoId.`)
+    }
+  } catch (err) {
+    throw makeApiClientError(400, err.message)
+  }
+
+  const responsePaths: FilesGetResponsePaths = {}
+
+  // Use max page size config to limit the paths processed
+  const paths = Object.keys(body.paths).slice(0, MAX_PAGE_SIZE)
```

Comment:
Should simply send an error if there are more than Max page size keys

Reply:
We should also add a new route to get the page size so that we could upgrade the page size on servers with no client side code changes

Reply:
Two ideas: `GET /maxPageSize` or more general purpose `GET /config`